### ISSUE TITLE
Forgets user data

### ISSUE THREAD

**Issue by ghost** [ttyridal/masterpassword-firefox#24]
I keep running into this issue where my full name is completely wiped out, and I have to type in my full name, master password and then re import my exported copy of the sites.

2.0.1pre2

**Comment by ttyridal** [comment]
You're sure you don't have private-browsing turned on?  Either as preferences->privacy->History->firefox will "never remember history" or otherwise?

masterpassword will write "won't store anything for private windows" to it's log if that's the case.

**Comment by ghost** [comment]
I don't believe I have any settings marked that would randomly clear the user data from this extension, but here are my privacy settings, and I don't usually use private mode.

https://i.imgur.com/guNgHfJ.png

**Comment by ttyridal** [comment]
No, that looks quite normal.. 

Do you see anything interesting in the log (Ctrl + Shift + J, filter on masterpassword)  when this happens?

if you are concerned about revealing your site settings skip lines with calc password xxxx . <index> . <passwordtype>

**Comment by ghost** [comment]
I will have to wait for it to happen again to check the console log.

**Comment by ghost** [comment]
I'm still waiting for it to happen again, but I did notice the following messages

Successfully compiled asm.js code (total compilation time 66ms; unable to cache asm.js in synchronous scripts; try loading asm.js via <script async> or createElement('script')) scrypt-asm.js

TypeError: asm.js type error: Disabled by debugger scrypt-asm.js

**Comment by ghost** [comment]
It happened a few minutes ago, but nothing showed in the console log.

**Comment by ttyridal** [comment]
Feels like this is going to be an interesting one. 
Keep an eye open for new beta releases. I'll probably need to simply add more logging as a first stage.

The asm.js warning should be nothing to care about. doing as they suggest opens up for a variety of timing / synchronisation issues (issue #8).

**Comment by ghost** [comment]
I've switched over to the beta channel on the Mozilla Addon Store

**Comment by ttyridal** [comment]
@linwiz one detail by the way. Does this happen mid-firefox-session, when starting a fresh firefox instance or a mix?

**Comment by ghost** [comment]
It seems to happen during a session.

**Comment by ttyridal** [comment]
Ok, thanks. That will refocus my strategy a bit I think.  

If you're lucky enough to not have any important tabs open when it happens, could you just restart Firefox to see if your config get's reloaded correctly from file?
masterpassword will read it's config only once, and then just keep it in memory.  The theory is that the addon is flushed somehow. (back to google...)

Do you run any memory cleanup plugins (RamBack, Memory Fox, Free Memory etc)

**Comment by ghost** [comment]
I will try that  and report next time it happens. I don't run anything of that sort, additionally I use Ubuntu 64bit w/ 8GB ram.

**Comment by ghost** [comment]
I decided to completely wipe my Firefox installation, remove all previous settings extensions etc. 
With a fresh start I can consistently get it to forget the data permanently by logging out of my computer or restarting it without closing Firefox first.

The extension also appears to have a problem when creating the user and importing a mpsites file. It doesn't seem to recognise that the config page has stored sites, instead it generates new sites based on the default long setting, not retrieving the stored sites. Restarting Firefox fixes this however, but it forgets the import and I am required to import one last time.

**Comment by ttyridal** [comment]
Hm. Your first case I can reproduce by force-closing (ie ctrl-c) firefox or logout from the x session, which seems to be the same thing. The latter is a Firefox bug in my opinion. (edit: and documented on bugzilla.mozilla.org)
The behaviour of masterpassword is in this case easily explained. 

Masterpassword is using a firefox api called sdk/simple-storage for storage.
 Firefox is obviously not flushing the storage right away (can easily be verified by observing that `<profiledir>/jetpack/jid1-pn4AFskf9WBAdA\@jetpack/simple-storage/` is empty).
Observing firefox for a little while, it seems the simple-storage is written to disk at clean shutdown and after a (rather long) timeout. (edit: this also has several entries on bugzilla.mozilla.org)

In other words, everything seems to be in order if Firefox is shutdown cleanly (ie Ctrl-Q or 'X' close window button) can you confirm that?

Could you elaborate on the second case?  If I start an empty firefox session (firefox --profile /tmp/emptydir), installs the addon, enter username and master key, open config, drop .mpsites, navigate to some site from .mpsites and clicks the masterpassword icon. I get the correct password generated. if I force-quit as above, all is lost, but if quitting firefox with Ctrl-Q or the 'X' - close window button, everything is in place when starting firefox anew.

This was with the latest 2.0.1pre3.

Is all your troubles related to "unclean" shutdown of Firefox?

**Comment by ghost** [comment]
Despite testing the second issue I just reported, I can not reproduce it again.

I believe that you are correct, it is due to unclean shutdown now, but originally it was something else--I believe you have addressed the original issue, but I will continue to keep an eye on that.

If the unclean shutdown is actually a firefox issue, then I guess it can be ignored.

**Comment by ttyridal** [comment]
I'll leave this open for a while in case you (or others) experiences similar problems. 

There seems to be a way in which we can save the settings without simple-storage's policies, and thus manage unclean shutdowns. [io_file api](https://developer.mozilla.org/en-US/Add-ons/SDK/Low-Level_APIs/io_file) (and this [stack overflow](http://stackoverflow.com/questions/23030084/how-to-save-a-textfile-with-firefox-add-on-sdk)). It's recommended to avoid the "low-level" api's, but we'll see.

**Comment by ghost** [comment]
I have not had any further issues, I would consider this resolved.
